stan/math/rev/fun/svd_U.hpp

```diff
@@ -0,0 +1,74 @@
+#ifndef STAN_MATH_REV_FUN_SVD_U_HPP
+#define STAN_MATH_REV_FUN_SVD_U_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err/check_symmetric.hpp>
+#include <stan/math/prim/err/check_nonzero_size.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <Eigen/SVD>
+
+namespace stan {
+namespace math {
+
+/**
+ * Given input matrix m, return matrix U where \mathbf{m} = \mathbf{UDV^{T}}
+ *
+ * This function internally calls Eigen::JacobiSVD to compute decomposition.
+ *
+ * Equation (4) from Differentiable Programming Tensor Networks(H. Liao, J. Liu, et al., arXiv:1903.09650)
+ * is used for MxN input matrix's adjoint calculation.
+ *
+ * @tparam EigMat type of input matrix
+ * @param m MxN input matrix
+ * @return Orthogonal matrix U
+ */
+template <typename EigMat,
+	  require_eigen_matrix_dynamic_t<EigMat>* = nullptr,
+	  require_vt_var<EigMat>* = nullptr>
+inline auto svd_U(const EigMat& m) {
+  using ret_type = promote_scalar_t<var, Eigen::MatrixXd>;
+  check_nonzero_size("singular_values", "m", m);
+
+  const int M = to_arena(m.rows());
+  auto arena_m = to_arena(m);
+  
+  Eigen::JacobiSVD<Eigen::MatrixXd> svd(arena_m.val(), Eigen::ComputeFullU | Eigen::ComputeFullV);
+
+  auto arena_D = to_arena((Eigen::MatrixXd) svd.singularValues().asDiagonal());
```

Comment:
This matrix isn't necessarily square, and if you uncomment the failing mixed mode tests I think this is causing a problem.

If the input is nonsquare, then this matrix has some extra zero rows or columns. You don't actually need to allocate a matrix with zero rows and columns --  you can just keep that in mine when you do operations with the matrix.

Check out how the autodiff in `singular_values` works. Instead of appending zeros to `D`, to compute `D * V^T`, we ignore the part of the matrix `V^T` that we'd multiply the zero stuff by.